### ISSUE TITLE
Write a Kubernetes event when the cluster is Ready

### ISSUE THREAD

**Issue by braghettos** [kubernetes-sigs/cluster-api#7424]
**User Story**

As an operator, I would like to get a Kubernetes event when the cluster is Ready. The use case is useful when I would like to trigger a notification or workflow, i.e. with Argo Events

**Detailed Description**

Actually there is already a Condition that shows when the cluster is Ready. It would be useful to add also a Kubernetes event stating the same thing.

**Anything else you would like to add:**

We have a condition for that on the cluster object:

// ControlPlaneInitializedCondition reports if the cluster's control plane has been initialized such that the
// cluster's apiserver is reachable and at least one control plane Machine has a node reference. Once this
// condition is marked true, its value is never changed. See the ControlPlaneReady condition for an indication of
// the current readiness of the cluster's control plane.
ControlPlaneInitializedCondition ConditionType = "ControlPlaneInitialized"

/kind feature

**Comment by fabriziopandini** [comment]
This seems a nice idea to me.
But before proceeding we should carefully identify what cluster ready is because there are different notions of it; e.g the condition about reports if the cluster's control plane has been initialized such that the cluster's apiserver is reachable and at least one control plane Machine has a node reference.

Is that what you mean by Cluster is ready?

/triage accepted
/hellp

**Comment by braghettos** [comment]
I think a Cluster is Ready from a user perspective when pods can be scheduled (no taints NoSchedule on worker nodes). I think it's aligned to what you are suggesting.

**Comment by braghettos** [comment]
+1 for your proposal @mtougeron

**Comment by braghettos** [comment]
Any updates on this opportunity?

**Comment by fabriziopandini** [comment]
> I think a Cluster is Ready from a user perspective when pods can be scheduled (no taints NoSchedule on worker nodes). I think it's aligned to what you are suggesting.

not sure I grasp the no taints NoSchedule part of it, the taints are usually managed by kubelet and CPI on a node bases, not by Cluster CAPI

> I think there are several different events that can be triggered here from the base CAPI perspective as well as the individual provider perspectives (e.g., when the managed control plane is available, when the network components are available, etc)

If I can give my 2 cents, in order to be actionable this should be more specif/potentially split in different issues

Assuming above topics could be addressed quickly
/help

**Comment by braghettos** [comment]
Regarding the https://cluster-api.sigs.k8s.io/tasks/experimental-features/runtime-sdk/implement-lifecycle-hooks.html, I think there's should be an additional state after 'API Server is available' when I have deployed a CNI and all the worker nodes are Ready.
Every event of this lifecycle should write a Kubernetes events.

**Comment by chrischdi** [comment]
I have some concerns about relying on an "exposed Kubernetes event", due to the eventual consistency of kubernetes/etcd and events getting deleted/garbage collected after some time, in my opinion it makes more sense to favor a "I have a way to check if this is the case by reading some related CR" (whatever I want to check, maybe best to simplify via some public library / functions).

If I get argo events right, they also don't rely on actual kubernetes events, they rely on kubernetes resources and events for them (ADD, UPDATE, DELETE, which exist for every k8s resource out of the box, but they are not an actual k8s resource) plus filtering on the object itself (labels, field values), but I might be wrong.

**Comment by mtougeron** [comment]
I agree it isn't a perfect solution but imo the ability to set a WATCH for these events is better than having nothing or having to constantly poll the apiserver for the status.

**Comment by braghettos** [comment]
Ok so we should open another issue regarding the opportunity to write a Kubernetes event for every new condition for every Custom Resource - but originally this issue was open by myself because we're missing the opportunity to write a Kubernetes event "when the cluster is Ready", which IMHO should translate into the possibility to schedule pods on worker nodes. Is there already a condition representing this state?

**Comment by mtougeron** [comment]
> but originally this issue was open by myself because we're missing the opportunity to write a Kubernetes event

I think this is a great opportunity to write these events. I'm fully supportive of it and if, when I get a moment, no one else has picked it up, I'd love to work on it.

>  "when the cluster is Ready"

My 2c is that I think you've put forward a reasonable definition of `Ready` from a CAPI point of view but I also agree with @fabriziopandini that some of it might be dependent on the individual providers; especially when it comes to managed control planes.

> Ok so we should open another issue

I don't have an opinion either way on this. I think it might be over complicating things unless the definition of `Ready` gets too dependent on the individual providers.

**Comment by sbueringer** [comment]
I think Ready could be super-ambigious. Are we talking about:
* when a cluster is "ready" after initial provisioning
* when a cluster is "ready" after latest changes to KCP/ MachineDeployments/ ... are reconciled + nodes are up ...

What should happen if Nodes & Machines and thus the Cluster flips into not ready? Are we expecting an event every time the cluster flips into not Ready / Ready?

**Comment by braghettos** [comment]
In the CAPI context, this is the status of a Cluster custom resource:

```
status:
  conditions:
  - lastTransitionTime: "2022-11-03T10:03:37Z"
    status: "True"
    type: Ready
  - lastTransitionTime: "2022-11-03T09:56:32Z"
    status: "True"
    type: ControlPlaneInitialized
  - lastTransitionTime: "2022-11-03T10:03:37Z"
    status: "True"
    type: ControlPlaneReady
  - lastTransitionTime: "2022-11-03T09:53:04Z"
    status: "True"
    type: InfrastructureReady
  controlPlaneReady: true
  failureDomains:
    az1:
      controlPlane: true
    az2:
      controlPlane: true
    nova:
      controlPlane: true
  infrastructureReady: true
  observedGeneration: 2
  phase: Provisioned
```

So I would say that the Cluster is Ready when both `controlPlaneReady` and `infrastructureReady` are `true`, what do you think?

There is also an event with reason `ControlPlaneUnhealthy`, why there shouldn't be one for ControlPlaneReady as the condition states?

> background information: every Kubernetes event in Kubernetes uses an etcd lease and they are refreshed when the etcd leader changes
That's clear, an event should be captured and stored if someone requires persistency of that event.

**Comment by mtougeron** [comment]
> So I would say that the Cluster is Ready when both controlPlaneReady and infrastructureReady are true, what do you think?

I would agree with this statement.


But while there still seems to be a lot of back and forth for what defines when the overall `Cluster` could be considered Ready from an event perspective what about simply emitting events according to some of the definitions from https://github.com/kubernetes-sigs/cluster-api/blob/main/docs/proposals/20200506-conditions.md?

I think we should come to a consensus on what a Ready `Cluster` resource is but for anything else we should defer to concepts from that proposal. Thoughts?

**Comment by braghettos** [comment]
@mtougeron let's organize a voting session / poll to define when the Cluster resource is ready - @fabriziopandini what do you think?

**Comment by mtougeron** [comment]
A voting session sounds like a good idea. I'll add this to the agenda for the next Dec 7th weekly session too.

**Comment by fabriziopandini** [comment]
I don't have objections to whatever method you will use to get an agreement.
However, It seems to me that an easier and less opinionated approach is to echo with events when existing conditions are transitioning, this will keep things nicely consistent without introducing a new definition of ready that can easily lead to discussions

**Comment by braghettos** [comment]
@mtougeron let me know if I can help in any way.

**Comment by mtougeron** [comment]
@braghettos sorry, been a hectic week for me. If you want to setup the voting session, go for it. Otherwise I'll try to pick this up Tue/Wed of next week. I'll check in with you via slack as well before I do anything.

**Comment by mtougeron** [comment]
We talked about this in today's community session. [Notes are here](https://docs.google.com/document/d/1ushaVqAKYnZ2VN_aa3GyKlS4kEd6bSug13xaXOakAQI/edit#). The tldr is that we felt that it'd be best to start simple with emitting events based on the status condition transitions that happen. We also need to write some additional, more detailed user stories around the use-cases for this functionality. I have an action item to do this for the Adobe use-cases.

/assign tougeron

**Comment by braghettos** [comment]
> We talked about this in today's community session. [Notes are here](https://docs.google.com/document/d/1ushaVqAKYnZ2VN_aa3GyKlS4kEd6bSug13xaXOakAQI/edit#). The tldr is that we felt that it'd be best to start simple with emitting events based on the status condition transitions that happen. We also need to write some additional, more detailed user stories around the use-cases for this functionality. I have an action item to do this for the Adobe use-cases.
> 
> /assign mtougeron

That's really great! What could be the expected release of such implementation?

**Comment by mtougeron** [comment]
> That's really great! What could be the expected release of such implementation?

@braghettos I'm not sure. I'd like to work on it over the next two weeks but I don't know for sure if that'll happen due to the holidays. If not, I'll start on it in early Jan.

**Comment by mtougeron** [comment]
The use cases from my POV that I would be implementing when this was implemented:

I want to be able to receive Kubernetes Events when a Cluster `status.conditions` changes so that I can:
- Watch for events for when a Cluster has a status condition of `ControlPlaneReady` so that I can perform related actions to the Cluster creation such as applying additional manifests through a 3rd party system.
- Watch for events for when a Cluster is `InfrastructureReady` so that I can trigger a 3rd party external workflow or event to create additional cloud infrastructure.
- Watch for events for when a Cluster is `Ready` so that I can trigger a 3rd party event (such as Argo Events) or a workflow (such as Argo Workflows) when a cluster change is completed.
- Watch for events to know when a Cluster is going through updates so that I can trigger 3rd party events (such as Argo Events) to notify tenants on the cluster that maintenance is being performed in real-time.
- Audit the timeline of Cluster events through event logging via my logging infrastructure.

I want to be able to receive Kubernetes Events when a Cluster `status.phase` changes to that I can:
- Watch for events to know when a Cluster has become `Provisioned` so that I can perform related actions to the Cluster creation. For example, registering the Cluster into Argo CD or into my company's change management system.

**Comment by mtougeron** [comment]
Looking at the code for this, I've found a fairly easy way to emit events based on the `status.phase` changes so I'm going to start simple with that.

The `status.conditions` changes will be a bit more work as I've found that some providers are already emitting some events.

**Comment by mtougeron** [comment]
@braghettos Does this set of events (copied from the PR description) seem reasonable to you? It's not from all the conditions as I mentioned in my previous comment but this seemed like a good start to get feedback on. Thoughts?

```
$> kubectl get events --sort-by=.metadata.creationTimestamp
LAST SEEN   TYPE     REASON                  OBJECT               MESSAGE
17m         Normal   Provisioning            cluster/tge-aks      Cluster is Provisioning
16m         Normal   InfrastructureReady     cluster/tge-aks      Cluster InfrastructureReady is now true
13m         Normal   ControlPlaneReady       cluster/tge-aks      Cluster ControlPlaneReady is now true
12m         Normal   Provisioned             cluster/tge-aks      Cluster is Provisioned
6m23s       Normal   Deleting                cluster/tge-aks      Cluster is Deleting
11s         Normal   Deleted                 cluster/tge-aks      Cluster has been deleted
```

**Comment by braghettos** [comment]
@mtougeron in particular 'Provisioned' what does it mean from a CAPI point of view? Anyway they look promising!

**Comment by braghettos** [comment]
@mtougeron did you had the chance to move on with this implementation?

**Comment by mtougeron** [comment]
@braghettos I received some feedback on the PR last week that I need to adjust for. I plan on picking that back up next week.

**Comment by braghettos** [comment]
@mtougeron any updates on your PR?

**Comment by mtougeron** [comment]
@braghettos I'm working on the PR today. I'd suggest following the PR directly for the latest updates on it; I've been updating the status there.

**Comment by braghettos** [comment]
@mtougeron I'm sorry I haven't rally seen the PR. Thank you for your great work.

**Comment by mtougeron** [comment]
@braghettos no worries. Just wanted to let you know about it so you had the latest info available as it happens :)

**Comment by sbueringer** [comment]
Can this issue be closed now that #7786 has been merged?

**Comment by braghettos** [comment]
Sure @sbueringer!
